bundles/org.openhab.binding.heos/src/main/java/org/openhab/binding/heos/internal/discovery/HeosPlayerDiscovery.java

```diff
@@ -133,52 +145,123 @@ protected void startScan() {
         removedGroups();
     }
 
+    /**
+     * This method returns a {@code Map<String pid, HeosPlayer heosPlayer>} with
+     * all player found on the network after an connection to the system is
+     * established via a bridge.
+     *
+     * @return a HashMap with all HEOS Player in the network
+     */
+    private @Nullable Map<Integer, Player> getNewPlayers() {
+        try {
+            playerMapNew.clear();
+
+            for (Player player : bridge.getPlayers()) {
+                playerMapNew.put(player.playerId, player);
+                removedPlayerMap = comparePlayerMaps(playerMapNew, playerMapOld);
+                playerMapOld.clear();
+                playerMapOld.putAll(playerMapNew);
+            }
+            return playerMapNew;
+        } catch (IOException | Telnet.ReadException e) {
+            logger.debug("Failed getting/processing groups", e);
+            return null;
+        }
+    }
+
+    /**
+     * This method searches for all groups which are on the HEOS network
+     * and returns a {@code Map<String gid, HeosGroup heosGroup>}.
+     * Before calling this method a connection via a bridge has to be
+     * established
+     *
+     * @return a HashMap with all HEOS groups
+     * @throws Telnet.ReadException
+     * @throws IOException
+     */
+    private @Nullable Map<String, Group> getNewGroups() {
+        groupMapNew.clear();
+        removedGroupMap.clear();
+
+        try {
+            Group[] groups = bridge.getGroups();
+
+            if (groups.length == 0) {
+                removedGroupMap = compareGroupMaps(groupMapNew, groupMapOld);
+                groupMapOld.clear();
+                return groupMapNew;
+            }
+
+            for (Group group : groups) {
+                logger.debug("Found: Group {} with {} Players", group.name, group.players.size());
+                groupMapNew.put(HeosGroup.calculateGroupMemberHash(group), group);
+                removedGroupMap = compareGroupMaps(groupMapNew, groupMapOld);
+                groupMapOld.clear(); // clear the old map so that only the currently available groups are added in the
+                                     // next
+                // step.
+                groupMapOld.putAll(groupMapNew);
```

Comment:
I'm not sure this logic is correct... I suggest rewriting it.

Reply:
That part should not be inside the loop :-)